internal/sqlauthproxy/mysql.go

```
@@ -0,0 +1,130 @@
+	return true, nil
+}
+
+func (p *dummyProvider) GetCredential(username string) (password string, found bool, err error) {
+	return "", true, nil
+}
+
+func newMysqlHandler(c Config) (*mysqlHandler, error) {
+	upstreamAddress := net.JoinHostPort(c.Hostname, fmt.Sprintf("%d", c.Port))
+	var options []func(*client.Conn)
+	var awsCredentials aws.CredentialsProvider
+	if c.RdsIam {
+		cfg, err := config.LoadDefaultConfig(context.TODO())
+		if err != nil {
+			return nil, fmt.Errorf("failed to load aws config: %s", err)
+		}
+
+		awsCredentials = cfg.Credentials
+	}
+
+	if c.UpstreamTLS && c.DialerFunc == nil {
+		tlsConfig := &tls.Config{}
+
+		if c.UpstreamCAFile != "" {
+			caPool, err := cert.NewPool(c.UpstreamCAFile)
+			if err != nil {
+				return nil, fmt.Errorf("failed to load upstream CA: %s", err)
+			}
+			tlsConfig.RootCAs = caPool
+			tlsConfig.ServerName = c.Hostname
+		} else {
+			tlsConfig.InsecureSkipVerify = true
+		}
+
+		if c.UpstreamCertFile != "" && c.UpstreamKeyFile != "" {
+			cert, err := tls.LoadX509KeyPair(c.UpstreamCertFile, c.UpstreamKeyFile)
+			if err != nil {
+				return nil, fmt.Errorf("failed to load upstream cert: %s", err)
+			}
+			tlsConfig.Certificates = []tls.Certificate{cert}
+		} else if c.UpstreamCertFile != "" || c.UpstreamKeyFile != "" {
+			return nil, fmt.Errorf("upstream cert and key must both be provided")
+		}
+
+		options = append(options, func(c *client.Conn) { c.SetTLSConfig(tlsConfig) })
+	}
+
+	return &mysqlHandler{
+		Config:          c,
+		server:          server.NewDefaultServer(),
+		upstreamAddress: upstreamAddress,
+		awsCredentials:  awsCredentials,
+		options:         options,
+	}, nil
+}
+
+func (h mysqlHandler) handleClient(c net.Conn) {
+	defer c.Close()
+
+	serverHandler := &mysqlServerHandler{}
+	clientConn, err := server.NewCustomizedConn(c, h.server, &dummyProvider{}, serverHandler)
+	if err != nil {
+		log.Printf("sqlauthproxy: failed to accept connection: %s", err)
+		return
```

should we return this error?

---

This function is a go routine on its own, so there is nothing to return to.